api/common/sharing/utils.go

```diff
@@ -12,64 +12,231 @@ import (
+	return objectList, err
+}
+
+func filterInstancesBySelectors(ctx context.Context, repository storage.Repository, instances types.ObjectList, parameters map[string]gjson.Result, smaap bool, offeringID string) (*types.ServiceInstance, error) {
+	var err error
+	var filteredInstances types.ServiceInstances
+
+	planSelectorVal, planSelectorExists := parameters[instance_sharing.ReferencePlanNameSelector]
+	var selectorPlan *types.ServicePlan
+	if planSelectorExists && len(planSelectorVal.String()) > 0 {
+		selectorPlan, err = getSelectorPlan(ctx, repository, smaap, offeringID, planSelectorVal.String())
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	for i := 0; i < instances.Len(); i++ {
+		instance := instances.ItemAt(i).(*types.ServiceInstance)
+
+		selectorVal, exists := parameters[instance_sharing.ReferenceInstanceNameSelector]
+		if exists && len(selectorVal.String()) > 0 && !(instance.Name == selectorVal.String()) {
+			continue
+		}
+
+		if selectorPlan != nil && instance.ServicePlanID != selectorPlan.ID {
+			continue
+		}
+
+		selectorVal, exists = parameters[instance_sharing.ReferenceLabelSelector]
+		if exists {
+			selectorLabels := types.Labels{}
+			if err := util.BytesToObject([]byte(selectorVal.Raw), &selectorLabels); err != nil {
+				return nil, err
+			}
+			if len(selectorLabels) > 0 {
+				match, err := matchLabels(instance.Labels, selectorLabels)
+				if err != nil {
+					return nil, err
+				}
+				if !match {
+					continue
+				}
+			}
+		}
+
+		filteredInstances.Add(instance)
+		// only single result is accepted for selectors:
+		if filteredInstances.Len() > 1 {
+			return nil, util.HandleInstanceSharingError(util.ErrMultipleReferenceSelectorResults, "")
+		}
+	}
+	if filteredInstances.Len() == 0 {
+		return nil, util.HandleInstanceSharingError(util.ErrNoResultsForReferenceSelector, "")
+	}
+	return filteredInstances.ItemAt(0).(*types.ServiceInstance), nil
+}
 
-	return referencedInstanceID.String(), nil
+func getSelectorPlan(ctx context.Context, repository storage.Repository, smaap bool, offeringID, selectorValue string) (*types.ServicePlan, error) {
+	var err error
+	var key string
+	if smaap {
+		key = "name"
+	} else {
+		key = "catalog_name"
+	}
+	planObject, err := repository.Get(ctx, types.ServicePlanType,
+		query.ByField(query.EqualsOperator, key, selectorValue),
+		query.ByField(query.EqualsOperator, "service_offering_id", offeringID),
+	)
+	if err != nil {
+		return nil, util.HandleInstanceSharingError(util.ErrNoResultsForReferenceSelector, "")
+	}
+	selectorPlan := planObject.(*types.ServicePlan)
+	return selectorPlan, nil
+}
+
+func matchLabels(instanceLabels types.Labels, selectorLabels types.Labels) (bool, error) {
+	for selectorLabelKey, selectorLabelArray := range selectorLabels {
+		instanceLabelArray, exists := instanceLabels[selectorLabelKey]
+		if exists && instanceLabels != nil {
+			for _, selectorLabelVal := range selectorLabelArray {
```

Comment:
you mean
if exists && len(instanceLabelArray) >0
right?
you can just check 
if exists &&  !contains(instanceLabelArray, selectorLabelVal) {
    return false, nil
}

}